### PR TITLE
add some basic structure for quantizer, io and simd operators

### PR DESCRIPTION
**Pull request by LHT129**
- only header for quantizer and io
- fp32 quantizer as example

### CODE REVIEW

File: src/io/memory_io.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    inline bool
+    MultiReadImpl(uint8_t* datas, uint64_t* sizes, uint64_t* offsets, uint64_t count) const;
+
+    inline void
+    PrefetchImpl(uint64_t offset, uint64_t cacheLine = 64);
+
+private:
+    [[nodiscard]] inline bool
+    CheckValidOffset(uint64_t size) const {
```
Comment (inabao) [inline src/io/memory_io.h]: unify the naming convention; our private methods generally use the underscore naming style. (check_valid_offset)

File: src/simd/fp32_simd_test.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+// limitations under the License.
+
+#include "fp32_simd.h"
+
+#include "catch2/benchmark/catch_benchmark.hpp"
+#include "catch2/catch_test_macros.hpp"
+#include "fixtures.h"
+using namespace vsag;
+
+#define TEST_RECALL(Func)                                                              \
```
Comment (inabao) [inline src/simd/fp32_simd_test.cpp]: Renaming it to TEST_ACCURACY would be more accurate.

File: src/simd/generic.cpp [hunk 3]
```diff
@@ -58,4 +59,25 @@ PQDistanceFloat256(const void* single_dim_centers, float single_dim_val, void* r
...
+    return result;
+}
+float
+FP32ComputeL2Sqr(const float* query, const float* codes, uint64_t dim) {
+    float result = 0.;
+    for (uint64_t i = 0; i < dim; ++i) {
+        auto val = query[i] - codes[i];
+        result += val * val;
+    }
+    return result;
```
Comment (inabao) [inline src/simd/generic.cpp:79]: There is no sqrt in the result. So, it is better to rename it to "FP32ComputeL2".

File: src/quantization/quantizer.h [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    cast() const {
+        return static_cast<const T&>(*this);
+    }
+
+    friend T;
+
+private:
+    uint64_t dim_{0};
+
+    uint64_t codeSize_{0};
```
Comment (inabao) [inline src/quantization/quantizer.h:209]: unify the code style.

File: src/simd/generic.cpp (hunk 3 above)
Comment (LHT129) [reply src/simd/generic.cpp:79]: L2Sqr is the Square  of L2 norm

File: src/simd/fp32_simd_test.cpp (hunk 2 above)
Comment (LHT129) [reply src/simd/fp32_simd_test.cpp]: done

File: src/quantization/quantizer_test.h [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    quant.ReTrain(vecs.data(), count);
+
+    // Test EncodeOne & DecodeOne
+    auto idx = random() % count;
+    auto* codes = new uint8_t[quant.GetCodeSize()];
+    quant.EncodeOne(vecs.data() + idx * dim, codes);
+    auto* outVec = new float[dim];
+    quant.DecodeOne(codes, outVec);
+    for (int i = 0; i < dim; ++i) {
+        REQUIRE(std::abs(vecs[idx * dim + i] - outVec[i]) < error);
```
Comment (ShawnShawnYou) [inline src/quantization/quantizer_test.h:36]: should replace abs with fabs?

File: src/simd/avx.cpp [hunk 6]
```diff
@@ -155,4 +155,50 @@ PQDistanceAVXFloat256(const void* single_dim_centers, float single_dim_val, void
...
+#if defined(ENABLE_AVX2)
+    const int n = dim / 8;             // process 8 floats at a time
+    __m256 sum = _mm256_setzero_ps();  // initialize to 0
+    for (int i = 0; i < n; ++i) {
+        __m256 a = _mm256_loadu_ps(query + i * 8);      // load 8 floats from memory
+        __m256 b = _mm256_loadu_ps(codes + i * 8);      // load 8 floats from memory
+        sum = _mm256_add_ps(sum, _mm256_mul_ps(a, b));  // accumulate the product
+    }
+    alignas(32) float result[8];
+    _mm256_store_ps(result, sum);  // store the accumulated result into an array
```
Comment (ShawnShawnYou) [inline src/simd/avx.cpp:174]: When n == 0, the store instruction (e.g. _mm256_store_ps) still occur in the avx512, avx2, and sse operators that may be harmful to the performance. Is it possible to add a judgment to skip these store instructions?

File: src/quantization/quantizer_test.h [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        float gt = 0.;
+        float value = 0.;
+        computer->ComputeDist(codes1, &value);
+        if (metric == vsag::MetricType::METRIC_TYPE_IP ||
+            metric == vsag::MetricType::METRIC_TYPE_COSINE) {
+            gt = InnerProduct(data + idx1 * dim, query + i * dim, &dim);
+        } else if (metric == vsag::MetricType::METRIC_TYPE_L2SQR) {
+            gt = L2Sqr(data + idx1 * dim, query + i * dim, &dim);
+        }
+        REQUIRE(std::abs(gt - value) < 1e-4);
```
Comment (ShawnShawnYou) [inline src/quantization/quantizer_test.h:192]: fabs?

File: src/simd/avx.cpp (hunk 6 above)
Comment (LHT129) [reply src/simd/avx.cpp:174]: yeah, I have fixed it

File: src/quantization/quantizer_test.h (hunk 5 above)
Comment (LHT129) [reply src/quantization/quantizer_test.h:36]: For integral arguments, the integral overloads of std::abs are likely better matches. 
reference: https://en.cppreference.com/w/cpp/numeric/math/fabs

**Review by inabao — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

File: src/io/basic_io.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+
+namespace vsag {
+
+template <typename IOTmpl>
```
Comment (jiaweizone) [inline src/io/basic_io.h:22]: `IOTmpl` need any interface limit ?

File: src/io/memory_io.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+MemoryIO::MultiReadImpl(uint8_t* datas, uint64_t* sizes, uint64_t* offsets, uint64_t count) const {
+    bool ret = true;
+    for (uint64_t i = 0; i < count; ++i) {
+        ret &= this->ReadImpl(datas, sizes[i], offsets[i]);
+        datas += sizes[i];
+    }
+    return ret;
+}
+void
+MemoryIO::PrefetchImpl(uint64_t offset, uint64_t cacheLine) {
```
Comment (jiaweizone) [inline src/io/memory_io.h:109]: explicit mark `cacheLine` to unused ?
cacheLine -> cache_line

File: src/quantization/fp32_quantizer.h [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+#pragma once
+#include <cstdint>
+#include <cstring>
+
+#include "quantizer.h"
+#include "simd/simd.h"
+
+namespace vsag {
+
+template <MetricType Metric = MetricType::METRIC_TYPE_L2SQR>
```
Comment (jiaweizone) [inline src/quantization/fp32_quantizer.h:25]: Metric -> metric

File: src/io/basic_io.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        return cast().ReadImpl(size, offset);
+    }
+
+    inline bool
+    MultiRead(uint8_t* datas, uint64_t* sizes, uint64_t* offsets, uint64_t count) const {
+        return cast().MultiReadImpl(datas, sizes, offsets, count);
+    }
+
+    inline void
+    Prefetch(uint64_t offset, uint64_t cacheLine = 64) {
```
Comment (jiaweizone) [inline src/io/basic_io.h:50]: cacheLine -> cache_line

File: src/io/memory_io.h [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    ReadImpl(uint8_t* data, uint64_t size, uint64_t offset) const;
+
+    [[nodiscard]] inline const uint8_t*
+    ReadImpl(uint64_t size, uint64_t offset) const;
+
+    inline bool
+    MultiReadImpl(uint8_t* datas, uint64_t* sizes, uint64_t* offsets, uint64_t count) const;
+
+    inline void
+    PrefetchImpl(uint64_t offset, uint64_t cacheLine = 64);
```
Comment (jiaweizone) [inline src/io/memory_io.h:51]: cacheLine -> cache_line

File: src/quantization/quantizer.h [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+    bool isTrained_{false};
```
Comment (jiaweizone) [inline src/quantization/quantizer.h:211]: isTrained_ -> is_trained_

File: src/quantization/quantizer.h (hunk 4 above)
Comment (jiaweizone) [reply src/quantization/quantizer.h:209]: codeSize -> code_size_

File: src/quantization/fp32_quantizer.h [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+float
+FP32Quantizer<Metric>::ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) {
+    if (Metric == MetricType::METRIC_TYPE_IP) {
+        return InnerProduct(codes1, codes2, &this->dim_);
+    } else if (Metric == MetricType::METRIC_TYPE_L2SQR) {
+        return L2Sqr(codes1, codes2, &this->dim_);
+    } else if (Metric == MetricType::METRIC_TYPE_COSINE) {
+        return InnerProduct(codes1, codes2, &this->dim_);  // TODO
+    } else {
+        return 0.;
```
Comment (jiaweizone) [inline src/quantization/fp32_quantizer.h:113]: 0.0 ?

File: src/io/memory_io.h (hunk 9 above)
Comment (LHT129) [reply src/io/memory_io.h:109]: code style will be modified in following pull request

**Pull request merged**